Sources/Signal.swift

```diff
@@ -51,96 +55,55 @@ public final class Signal<Value, Error: Swift.Error> {
 		/// Disposing of `disposable` is assumed to remove the generator
 		/// observer from its attached `Signal`, so that the generator observer
 		/// as the last +1 retain of the `Signal` core may deinitialize.
-		private let disposable: CompositeDisposable
-
-		/// The state of the signal.
-		private var state: State
+		fileprivate let disposable: CompositeDisposable
 
-		/// Used to ensure that all state accesses are serialized.
+		/// Lock for states.
 		private let stateLock: Lock
 
-		/// Used to ensure that events are serialized during delivery to observers.
-		private let sendLock: Lock
+		/// The observers of the `Signal`.
+		private var observers: Bag<Observer>?
 
-		fileprivate init(_ generator: (Observer, Lifetime) -> Void) {
-			state = .alive(Bag(), hasDeinitialized: false)
+		/// Whether the `Signal` has been deinitialized and hence no any further
+		/// observer can ever be attached.
+		private var hasDeinitialized: Bool
 
-			stateLock = Lock.make()
-			sendLock = Lock.make()
+		fileprivate init(_ generator: (Observer, Lifetime) -> Void) {
 			disposable = CompositeDisposable()
+			stateLock = Lock.make()
+			observers = Bag()
+			hasDeinitialized = false
+
+			disposable += AnyDisposable { [weak self] in
+				guard let strongSelf = self else { return }
+
+				/// Ensure the observers are deinitialized after we release the
+				/// lock, since an observer could eventually result in
+				/// an synchronous attempt to detach observers from `self`.
+				withExtendedLifetime(strongSelf.observers) {
+					strongSelf.stateLock.lock()
+					strongSelf.observers = nil
+					strongSelf.stateLock.unlock()
+				}
+			}
 
 			// The generator observer retains the `Signal` core.
-			generator(Observer(action: self.send, interruptsOnDeinit: true), Lifetime(disposable))
+			let sink = Event.makeSynchronizing(self.send, disposable: disposable)
+			let inputObserver = Observer(action: sink, interruptsOnDeinit: true)
+			generator(inputObserver, Lifetime(disposable))
 		}
 
 		private func send(_ event: Event) {
```

Comment:
I think we should remove this function, inlining it into the `init`. It's no longer safe to call this outside of `sink` above, so leaving it here risks future violations of the signal contract.